File: Company 2 : Amazon/Day 3/8. Count ways to N'th Stair/code.cpp
class Solution {
public:
    long long nCr(int n, int r) {
        r = max(r, n - r);
        long long ans = 1;
        for (int i = 1; i <= r; i++) {
            ans *= (n - r + i);
            ans /= i;
        }
        return ans;
    }
    int waysToReachStair(int k) {
        if (k == 0 || k == 4)
            return 2;
        else if (k == 1 || k == 2)
            return 4;
        else {
            long long r = pow(2, ceil(log2(k))) - k;
            if (r < 0 || r > ceil(log2(k)) + 1)
                return 0;
            else
                return nCr(ceil(log2(k)) + 1, r);
        }
    }
};

File: Company 2 : Amazon/Day 3/8. Count ways to N'th Stair/code.java
class Solution {
    public int comb(int n, int k) {
        if (k < 0 || k > n)
            return 0;
        long res = 1;
        for (int i = 0; i < k; i++) {
            res = res * (n - i) / (i + 1);
        }
        return (int) res;
    }

    public int waysToReachStair(int k) {
        int res = 0;
        for (int i = 0; i < 30; i++) {
            res += comb(i + 1, (1 << i) - k);
        }
        return res;
    }
}
